File: core/src/commonMain/kotlin/work/socialhub/kmastodon/MastodonFactory.java
package work.socialhub.kmastodon;

import mastodon4j.domain.Service;
import mastodon4j.internal._Mastodon;

/**
 * @author hecateball
 */
public final class MastodonFactory {

    private MastodonFactory() {
    }

    /**
     * get request instance without service.
     */
    public static Mastodon getInstance(String uri, String accessToken) {
        return new _Mastodon(null, uri, accessToken);
    }

    /**
     * get request instance with service.
     */
    public static Mastodon getInstance(Service service, String uri, String accessToken) {
        return new _Mastodon(service, uri, accessToken);
    }
}

File: core/src/commonMain/kotlin/work/socialhub/kmastodon/MastodonFactory.kt
package work.socialhub.kmastodon

import work.socialhub.kmastodon.domain.Service
import work.socialhub.kmastodon.internal.MastodonImpl
import work.socialhub.kmastodon.internal.PixelfedImpl
import work.socialhub.kmastodon.internal.PleromaImpl
import kotlin.js.JsExport

@JsExport
object MastodonFactory {

    /**
     * get request instance
     *
     * Pass `service = null` to auto-detect the instance software
     * (Mastodon, Pleroma, Akkoma, Pixelfed, GoToSocial, ...) via NodeInfo
     * on the first request. Defaults to [Service.MASTODON] for compatibility.
     */
    fun instance(
        uri: String,
        accessToken: String = "",
        service: Service? = Service.MASTODON,
    ): Mastodon {
        return MastodonImpl(uri, accessToken, service)
    }

    /**
     * get a Pleroma / Akkoma client, exposing Pleroma-specific resources
     * (emoji reactions, chats, ...) in addition to the Mastodon API.
     */
    fun pleroma(
        uri: String,
        accessToken: String = "",
        service: Service? = Service.PLEROMA,
    ): Pleroma {
        return PleromaImpl(uri, accessToken, service)
    }

    /**
     * get a Pixelfed client, exposing Pixelfed-specific resources
     * (stories, collections) in addition to the Mastodon API.
     */
    fun pixelfed(
        uri: String,
        accessToken: String = "",
        service: Service? = Service.PIXELFED,
    ): Pixelfed {
        return PixelfedImpl(uri, accessToken, service)
    }
}
